application-licensing-licensor/application-licensing-licensor-api/src/main/java/com/xwiki/licensing/internal/DefaultLicensedExtensionManager.java

```diff
@@ -116,23 +121,69 @@
     }
 
     @Override
-    public Collection<ExtensionId> getMandatoryLicensedExtensions()
+    public Set<ExtensionId> getMandatoryLicensedExtensions()
+    {
+        Set<ExtensionId> mandatoryLicensedExtensions = computeMandatoryLicensedExtensions();
+        this.cachedMandatoryLicensedExtensions = mandatoryLicensedExtensions;
+        return mandatoryLicensedExtensions;
+    }
+
+    private synchronized Set<ExtensionId> computeMandatoryLicensedExtensions()
+    {
+        if (this.cachedMandatoryLicensedExtensions != null) {
+            return this.cachedMandatoryLicensedExtensions;
+        }
```

Comment:
The risk is that some other thread can set the ``cachedMandatoryLicensedExtensions`` to ``null`` between the test and the return.

```suggestion
        Set<ExtensionId> mandatoryLicensedExtensions = this.cachedMandatoryLicensedExtensions;
        if (mandatoryLicensedExtensions != null) {
            return mandatoryLicensedExtensions;
        }
```